lxd/instance/instance_utils.go

```diff
@@ -1364,3 +1364,108 @@ func PlanAttachedVolumesForSnapshot(s *state.State, inst Instance, diskVolumesMo
 
 	return attached, nil
 }
+
+// ResolveAttachedVolumeSnapshotsForRestore resolves the custom storage volume
+// snapshots that were taken alongside sourceSnapshot. It reads the snapshot's
+// "volatile.attached_volumes" (JSON map of volumeUUID -> snapshotUUID) and
+// returns the matching snapshot volumes. Fails if a referenced snapshot is
+// missing. When diskVolumesMode is "all-exclusive", fails if any of the source
+// volumes are now shared.
+func ResolveAttachedVolumeSnapshotsForRestore(s *state.State, sourceSnapshot Instance, diskVolumesMode string) (attached []*api.StorageVolume, err error) {
+	allExclusiveMode := false
+	switch diskVolumesMode {
+	case api.DiskVolumesModeRoot, "":
+		// Root disk only, nothing to do.
+		return nil, nil
+	case api.DiskVolumesModeAllExclusive:
+		allExclusiveMode = true
+	default:
+		return nil, errors.New("Invalid disk volumes mode")
+	}
+
+	raw, ok := sourceSnapshot.LocalConfig()["volatile.attached_volumes"]
+	if !ok {
+		return nil, errors.New(`Cannot get attached volumes as "volatile.attached_volumes" is unset`)
+	}
+
+	var snapByVolUUID map[string]string
+	err = json.Unmarshal([]byte(raw), &snapByVolUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get custom storage volumes in the instance's effective project.
+	volumes, err := getProjectVolumes(s, sourceSnapshot)
```

Comment:
Again, here, really this function should only be retrieving the volumes referenced by the sourceSnapshot instance.

So perhaps this function should be re-factored to parse the supplied instance's devices and load only the referenced volumes.